File: Design-Patterns-Using-CSharp/Decorator/DecoratorPatternWindow.cs
using System;

// Window (Component)
interface IWindow
{
    void Draw();
}

// SimpleWindow (ConcreteComponent)
class SimpleWindow : IWindow
{
    public void Draw()
    {
        Console.WriteLine("SimpleWindow draw.");
    }
}

// Decorator (Decorator)
abstract class Decorator : IWindow
{
    protected IWindow Component;

    public Decorator(IWindow component)
    {
        Component = component;
    }

    public virtual void Draw()
    {
        Component.Draw();
    }
}

// VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator : Decorator
{
    public VerticalScrollBarDecorator(IWindow component) : base(component)
    {
    }

    public override void Draw()
    {
        base.Draw();
        Console.WriteLine("VerticalScrollBarDecorator draw");
    }
}

// HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator : Decorator
{
    public HorizontalScrollBarDecorator(IWindow component) : base(component)
    {
    }

    public override void Draw()
    {
        base.Draw();
        Console.WriteLine("HorizontalScrollBarDecorator draw");
    }
}

// Client code
class DecoratorPatternWindow
{
    static void Main(string[] args)
    {
        IWindow component = new SimpleWindow();
        IWindow decorator1 = new VerticalScrollBarDecorator(component);
        IWindow decorator2 = new HorizontalScrollBarDecorator(decorator1);
        decorator2.Draw();
    }
}

/*
SimpleWindow draw.
VerticalScrollBarDecorator draw
HorizontalScrollBarDecorator draw
*/

File: Design-Patterns-Using-Ruby/Decorator/DecoratorPatternWindow.rb
# Window (Component)
class Window
    def draw
    end
end

# SimpleWindow (ConcreteComponent)
class SimpleWindow < Window
    def draw
        puts 'SimpleWindow draw.'
    end
end

# Decorator (Decorator)
class Decorator < Window
    attr_reader :component

    def initialize(component)
        @component = component
    end

    def draw
        component.draw
    end
end

# VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator < Decorator
    def initialize(component)
        super(component)
    end

    def draw
        super
        puts 'VerticalScrollBarDecorator draw'
    end
end

# HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator < Decorator
    def initialize(component)
        super(component)
    end

    def draw
        super
        puts 'HorizontalScrollBarDecorator draw'
    end
end

# Client code
component = SimpleWindow.new
decorator1 = VerticalScrollBarDecorator.new(component)
decorator2 = HorizontalScrollBarDecorator.new(decorator1)
decorator2.draw

=begin 
SimpleWindow draw.
VerticalScrollBarDecorator draw
HorizontalScrollBarDecorator draw 
=end